pkg/engine/lifecycletest/fuzzing/snapshot.go

```diff
@@ -0,0 +1,323 @@
+	ResourceCount: rapid.IntRange(2, 5),
+	Action:        rapid.SampledFrom(snapshotSpecActions),
+	ResourceOpts:  defaultResourceSpecOptions,
+}
+
+var snapshotSpecActions = []SnapshotSpecAction{
+	SnapshotSpecNew,
+	SnapshotSpecOld,
+	SnapshotSpecProvider,
+}
+
+// Given a set of StackSpecOptions and SnapshotSpecOptions, returns a rapid.Generator that yields random SnapshotSpecs.
+func GeneratedSnapshotSpec(sso StackSpecOptions, snso SnapshotSpecOptions) *rapid.Generator[*SnapshotSpec] {
+	sso = defaultStackSpecOptions.With(sso)
+	snso = defaultSnapshotSpecOptions.With(snso)
+
+	return rapid.Custom(func(t *rapid.T) *SnapshotSpec {
+		ss := &SnapshotSpec{}
+
+		newResource := generatedNewResourceSpec(ss, sso, snso.ResourceOpts)
+		oldResource := generatedOldResourceSpec(ss, sso, snso.ResourceOpts)
+		providerResource := generatedProviderResourceSpec(ss, sso)
+
+		// Seed the snapshot with an initial provider resource that can be used.
+		initialProvider := providerResource.Draw(t, "SnapshotSpec.InitialProvider")
+		AddTag(initialProvider, "snapshot.provider.initial")
+		ss.AddProvider(initialProvider)
+
+		resourceCount := snso.ResourceCount.Draw(t, "SnapshotSpec.ResourceCount")
+
+		for i := 0; i < resourceCount; i++ {
+			action := snso.Action.Draw(t, "SnapshotSpec.Action")
+
+			switch action {
+			case SnapshotSpecNew:
+				r := newResource.Draw(t, fmt.Sprintf("SnapshotSpec.ResourceSpec[%d]", i))
+				ss.AddResource(r)
+			case SnapshotSpecOld:
+				r := oldResource.Draw(t, fmt.Sprintf("SnapshotSpec.ResourceSpec[%d]", i))
+				ss.AddResource(r)
+			case SnapshotSpecProvider:
+				r := providerResource.Draw(t, fmt.Sprintf("SnapshotSpec.ResourceSpec[%d]", i))
+				ss.AddProvider(r)
+			}
+		}
+
+		return ss
+	})
+}
+
+func generatedNewResourceSpec(
+	ss *SnapshotSpec,
+	sso StackSpecOptions,
+	rso ResourceSpecOptions,
+) *rapid.Generator[*ResourceSpec] {
+	return rapid.Custom(func(t *rapid.T) *ResourceSpec {
+		r := GeneratedResourceSpec(sso, rso, ss.Providers).Draw(t, "NewResourceSpec.Base")
+		rds := GeneratedResourceDependencies(ss, r, includeAll).Draw(t, "NewResourceSpec.Dependencies")
+
+		r.Dependencies = rds.Dependencies
+		r.PropertyDependencies = rds.PropertyDependencies
+		r.DeletedWith = rds.DeletedWith
+
+		return r
+	})
+}
+
+func generatedOldResourceSpec(
+	ss *SnapshotSpec,
+	sso StackSpecOptions,
+	rso ResourceSpecOptions,
+) *rapid.Generator[*ResourceSpec] {
+	return rapid.Custom(func(t *rapid.T) *ResourceSpec {
+		var r *ResourceSpec
+		if len(ss.Resources) == 0 {
+			r = GeneratedResourceSpec(sso, rso, ss.Providers).Draw(t, "OldResourceSpec.New")
+		} else {
+			r = rapid.SampledFrom(ss.Resources).Draw(t, "OldResourceSpec.Base").Copy()
+		}
+
+		AddTag(r, SnapshotSpecOld)
+		rds := GeneratedResourceDependencies(ss, r, isDeleted).Draw(t, "OldResourceSpec.Dependencies")
+
+		r.Delete = true
+		r.Protect = rapid.Bool().Draw(t, "OldResourceSpec.Protect")
+		r.RetainOnDelete = rapid.Bool().Draw(t, "OldResourceSpec.RetainOnDelete")
```

Comment:
Should these also be set on new resources?

Reply:
I'll add a comment to the code here to explain -- in essence, we either generate a brand new resource if there are none to copy, or we copy one from the existing snapshot. This resource will already have some random `Protect` and `RetainOnDelete` attributes. We then, as well as setting `Delete` to true, randomly change some of these attributes to simulate a case in which e.g. you are replacing an old resource that was protected with a new resource that isn't.